stdlib/public/core/EnumeratedSequence.swift

```diff
@@ -0,0 +1,128 @@
+
+/// An enumeration of the elements of a sequence or collection.
+///
+/// `EnumeratedSequence` is a sequence of pairs (*n*, *x*), where *n*s are
+/// consecutive `Int` values starting at zero, and *x*s are the elements of a
+/// base sequence.
+///
+/// To create an instance of `EnumeratedSequence`, call `enumerated()` on a
+/// sequence or collection. The following example enumerates the elements of
+/// an array.
+///
+///     var s = ["foo", "bar"].enumerated()
+///     for (n, x) in s {
+///         print("\(n): \(x)")
+///     }
+///     // Prints "0: foo"
+///     // Prints "1: bar"
+@frozen
+public struct EnumeratedSequence<Base: Sequence> {
+  @usableFromInline
+  internal var _base: Base
+
+  /// Construct from a `Base` sequence.
+  @inlinable
+  internal init(_base: Base) {
+    self._base = _base
+  }
+}
+
+extension EnumeratedSequence {
+  /// The iterator for `EnumeratedSequence`.
+  ///
+  /// An instance of this iterator wraps a base iterator and yields
+  /// successive `Int` values, starting at zero, along with the elements of the
+  /// underlying base iterator. The following example enumerates the elements of
+  /// an array:
+  ///
+  ///     var iterator = ["foo", "bar"].enumerated().makeIterator()
+  ///     iterator.next() // (0, "foo")
+  ///     iterator.next() // (1, "bar")
+  ///     iterator.next() // nil
+  ///
+  /// To create an instance, call
+  /// `enumerated().makeIterator()` on a sequence or collection.
+  @frozen
+  public struct Iterator {
+    @usableFromInline
+    internal var _base: Base.Iterator
+    @usableFromInline
+    internal var _count: Int
+
+    /// Construct from a `Base` iterator.
+    @inlinable
+    internal init(_base: Base.Iterator) {
+      self._base = _base
+      self._count = 0
+    }
+  }
+}
+
+extension EnumeratedSequence.Iterator: IteratorProtocol, Sequence {
+  /// The type of element returned by `next()`.
+  public typealias Element = (offset: Int, element: Base.Element)
+
+  /// Advances to the next element and returns it, or `nil` if no next element
+  /// exists.
+  ///
+  /// Once `nil` has been returned, all subsequent calls return `nil`.
+  @inlinable
+  public mutating func next() -> Element? {
+    guard let b = _base.next() else { return nil }
+    let result = (offset: _count, element: b)
+    _count += 1 
+    return result
+  }
+}
+
+extension EnumeratedSequence: Sequence {
+  /// Returns an iterator over the elements of this sequence.
+  @inlinable
+  public __consuming func makeIterator() -> Iterator {
+    return Iterator(_base: _base.makeIterator())
+  }
+}
+
+extension EnumeratedSequence: Collection where Base: Collection {
+  public subscript(position: Base.Index) -> (offset: Int, element: Base.Element) {
+    get {
+      // This could probably be more efficient for non random access collections
+      return (offset: distance(from: startIndex, to: position), element: _base[position])
```

Comment:
Unfortunately this won't work as an implementation strategy. For non-random-access collections, `distance(from: startIndex, to: position)` is a linear operation. So every time you call it, it will walk from the start index to the position – which would mean regular iteration using indices would take quadratic time. A better scheme would probably be to smuggle the count in the `Index` type.

Reply:
I just wrote up a much more thoughtful reply, but accidentally lost it all by going back in my browser, so I'm going to summarize my thoughts and can expand later if there are any questions :(

I just realized that this subscript function was suppose to guarantee `O(1)` runtime.

> A better scheme would probably be to smuggle the count in the `Index` type.

Were you thinking EnumeratedSequence would have it's own Index type that wrapped Base.Index and also included the count? If yes, I'm not sure when it would make sense to create this. If at init then `.enumerated()` becomes `O(n)` in all cases which doesn't seem like a great option. Only other thing I could think of is lazily computed, but then that breaks the `O(1)`.

Maybe EnumeratedSequence should only conform to Collection when Base conforms to RandomAccessCollection?